### PR TITLE
【Prim】Support softmax grad

### PR DESCRIPTION
**Pull request by JiabinYang**
### PR types

New features
### PR changes

OPs
### Describe

Pcard-66975
This PR support softmax custom vjp

### CODE REVIEW

File: paddle/fluid/prim/api/composite_backward/composite_backward_api.h [hunk 1]
```diff
@@ -24,13 +24,43 @@ using Tensor = paddle::Tensor;
...
+      } else {
+        auto new_out_grad = out_grad * out;
+        auto tmp_x_grad =
+            new_out_grad - out * sum<T>(new_out_grad,
+                                        {out.dims().size() + axis},
+                                        out.dtype(),
+                                        true);
+        set_output<T>(tmp_x_grad, x_grad);
+      }
+    } else {
```
Comment (xiaoguoguo626807) [inline paddle/fluid/prim/api/composite_backward/composite_backward_api.h:54]: how could out_grad.dims().size() <= 0?

File: paddle/phi/api/yaml/legacy_backward.yaml [hunk 2]
```diff
@@ -1189,6 +1189,7 @@
     param : [out]
   kernel :
     func : softmax_grad
+  composite : softmax_grad(out, out_grad, axis)
```
Comment (xiaoguoguo626807) [inline paddle/phi/api/yaml/legacy_backward.yaml:1194]: add x_grad

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_resnet_v2.py [hunk 3]
```diff
@@ -427,6 +427,8 @@ def test_resnet(self):
 
     def test_resnet_composite(self):
         core._set_prim_backward_enabled(True)
+        # TODO(jiabin): We partially support custom vjp, disable softmax backward make sure this doesn't work util custom vjp enabled
+        core._add_skip_comp_ops("softmax")
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_resnet_v2.py:431]: can it be successful if add softmax ?

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    inputs.stop_gradient = False
+    res = fn(inputs)
+
+    gradients = paddle.grad(res, inputs)
+    return gradients
+
+
+class TestCompositeSoftmax(unittest.TestCase):
+    def setUp(self):
+        self.dtypes = ["float32", "float64"]
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py:75]: add fp16

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            )
+            x.stop_gradient = False
+            y = fn(x)
+            blocks = main_program.blocks
+
+            fwd_ops = [op.type for op in blocks[0].ops]
+            # Ensure that softmax in original block
+            self.assertTrue('softmax' in fwd_ops)
+
+            paddle.incubate.autograd.to_prim(blocks)
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py]: it change to paddle.incubate.autograd.primapi.to_prim()

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            x = paddle.static.data(
+                'x', shape=inputs.shape, dtype=str(inputs.dtype)
+            )
+            x.stop_gradient = False
+            y = fn(x)
+            blocks = main_program.blocks
+            # skip softmax forward decompose
+            core._set_prim_forward_blacklist("softmax")
+            paddle.incubate.autograd.to_prim(blocks)
+            z = paddle.static.gradients([y], x)
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py:161]: only check soft_max vjp, it's better to check forward composite and vjp

File: paddle/fluid/prim/api/composite_backward/composite_backward_api.h (hunk 1 above)
Comment (JiabinYang) [reply paddle/fluid/prim/api/composite_backward/composite_backward_api.h:54]: 0-D

File: paddle/phi/api/yaml/legacy_backward.yaml (hunk 2 above)
Comment (JiabinYang) [reply paddle/phi/api/yaml/legacy_backward.yaml:1194]: But x_grad is output?

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_resnet_v2.py (hunk 3 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_resnet_v2.py:431]: random failed...

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py (hunk 4 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py:75]: Softmax api doesn't support fp16

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py (hunk 6 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py:161]: once custom vjp pr merged, I'll fix it in next pr

File: python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py (hunk 5 above)
Comment (JiabinYang) [reply python/paddle/fluid/tests/unittests/prim/composite_ops/test_composite_softmax_custom_vjp.py]: done

**Pull request closed without merging**